js/misc/missionGameService.js

```diff
@@ -0,0 +1,267 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+
+const Lang = imports.lang;
+const GLib = imports.gi.GLib;
+const Gio = imports.gi.Gio;
+const Showmehow = imports.gi.Showmehow;
+const Signals = imports.signals;
+
+
+/**
+ * callWhenComplete
+ *
+ * Takes an object with keys mapping to bool values, a callback
+ * and a key name. Flips the key name to true and checks if
+ * they are all true, and if so, calls the callback.
+ *
+ * This is used by MissionChatboxTextService to ensure that
+ * all asynchronous tasks are completed before calling
+ * the completion callback. The callback is called with
+ * the key that triggered the callback.
+ */
+function callWhenComplete(completion, key, callback) {
+    completion[key] = true;
+
+    /* This is a shorthand way of implementing Array.all
+     * using De Morgan's laws and Array.some */
+    const complete = !Object.keys(completion).some(function(key) {
+        return !completion[key];
+    });
+
+    if (complete && callback) {
+        callback.apply(this, [key].concat(Array.from(arguments)));
+    }
+}
+
+const MissionChatboxTextService = new Lang.Class({
+    Name: 'MissionChatboxTextService',
+    Extends: GLib.GObject,
```

Comment:
It emits signals, which is why I had it as a GObject. Is there a way to have normal JSObjects emit signals without being bound to a GObject instance?